src/psyclone/psyir/transformations/matmul2code_trans.py

```diff
@@ -0,0 +1,286 @@
+        specified node. This node must be a MATMUL
+        BinaryOperation. Currently only the matrix vector version of
+        MATMUL is supported. If the transformation is successful then
+        an assignment which includes a MATMUL BinaryOperation node
+
+        .. code-block:: fortran
+
+            R = MATMUL(A,B)
+
+        is converted to the following equivalent inline code:
+
+        .. code-block:: fortran
+
+            do i=1,N
+                R(i) = 0.0
+                    do j=1,M
+                        R(i) = R(i) + A(i,j) * B(j)
+
+        :param node: a MATMUL Binary-Operation node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(node)
+
+        assignment = node.parent
+        matrix = node.children[0]
+        vector = node.children[1]
+        result = node.parent.lhs
+        result_symbol = result.symbol
+
+        # Create new i and j loop iterators
+        symbol_table = node.find_symbol_table()
+        i_loop_name = symbol_table.new_symbol_name("i")
+        i_loop_symbol = DataSymbol(i_loop_name, DataType.INTEGER)
+        symbol_table.add(i_loop_symbol)
+        j_loop_name = symbol_table.new_symbol_name("j")
+        j_loop_symbol = DataSymbol(j_loop_name, DataType.INTEGER)
+        symbol_table.add(j_loop_symbol)
+
+        # Create "result(i)"
+        result_dims = [Reference(i_loop_symbol)]
+        if len(result.children) > 1:
+            # Add any additional dimensions (in case of an array slice)
+            result_dims.extend(result.children[1:])
+        result = Array.create(result_symbol, result_dims)
+        # Create "vector(j)"
+        vector_dims = [Reference(j_loop_symbol)]
+        if len(vector.children) > 1:
+            # Add any additional dimensions (in case of an array slice)
+            vector_dims.extend(vector.children[1:])
+        vector_array_reference = Array.create(
+            vector.symbol, vector_dims)
+        # Create "matrix(i,j)"
+        array_dims = [Reference(i_loop_symbol), Reference(j_loop_symbol)]
+        if len(matrix.children) > 2:
+            # Add any additional dimensions (in case of an array slice)
+            array_dims.extend(matrix.children[2:])
+        matrix_array_reference = Array.create(matrix.symbol, array_dims)
+        # Create "matrix(i,j) * vector(j)"
+        multiply = BinaryOperation.create(
+            BinaryOperation.Operator.MUL, matrix_array_reference,
+            vector_array_reference)
+        # Create "result(i) + matrix(i,j) * vector(j)"
+        rhs = BinaryOperation.create(
+            BinaryOperation.Operator.ADD, result, multiply)
+        # Create "result(i) = result(i) + matrix(i,j) * vector(j)"
+        assign = Assignment.create(result, rhs)
+        # Create j loop and add the above code as a child
+        jloop = Loop.create(
```

Comment:
I think that both of the `Loop.create()` calls here are wrong - the upper bound for each needs to be a `UBOUND()`, not the bare matrix or vector symbol.

Reply:
Oh, and the lower bound should be `LBOUND()`, not just 1.

Reply:
Well that was a bit embarrassing. I thought I was pulling in the size of the symbol extent, not the symbol name.

Reply:
I only partly agree about the use of LBOUND and UBOUND. In the current examples we know the extent of the arrays from their declaration so can use those values. However, I do need to extend the tests to examples where the extent is not known. In this case LBOUND and UBOUND would need to be used.

Reply:
OK, I think I've got this working.

Reply:
Great.